### ISSUE TITLE
Corrupt fs after tying to rsync some 90G on a 150G single fs pool

### ISSUE THREAD

**Issue by yaneti** [stratis-storage/stratisd#1242]
```
# uname -r
4.19.0-0.rc5.git3.2.fc30.x86_64
# rpm -q stratisd
stratisd-1.0.0-1.module_2238+3c8433ea.x86_64
# lvdisplay  d2/skoji
  --- Logical volume ---
  LV Path                /dev/d2/skoji
  LV Name                skoji
  VG Name                d2
  LV UUID                c376VK-9qaZ-a98p-W1eq-Jl07-bqto-6JJias
  LV Write Access        read/write
  LV Creation host, time d2, 2018-10-02 13:57:04 +0300
  LV Status              available
  # open                 0
  LV Size                150.00 GiB
  Current LE             4800
  Segments               1
  Allocation             inherit
  Read ahead sectors     auto
  - currently set to     256
  Block device           253:8
# stratis pool
Name    Total Physical Size  Total Physical Used
# stratis pool create stest /dev/d2/skoji
# stratis pool 
Name     Total Physical Size  Total Physical Used
stest                150 GiB               52 MiB
# stratis fs create stest skoji
# stratis fs 
Pool Name  Name   Used     Created            Device                    
stest      skoji  546 MiB  Oct 02 2018 17:45  /dev/stratis/stest/skoji 
# du -s  /koji
93725192	/koji
# rsync -a --stats  /koji/ /koji2/

Number of files: 29,555 (reg: 29,506, dir: 48, link: 1)
Number of created files: 29,554 (reg: 29,506, dir: 47, link: 1)
Number of deleted files: 0
Number of regular files transferred: 29,506
Total file size: 95,914,923,278 bytes
Total transferred file size: 95,914,923,248 bytes
Literal data: 95,914,923,248 bytes
Matched data: 0 bytes
File list size: 1,376,010
File list generation time: 0.050 seconds
File list transfer time: 0.000 seconds
Total bytes sent: 95,940,931,462
Total bytes received: 561,292

sent 95,940,931,462 bytes  received 561,292 bytes  116,930,521.33 bytes/sec
total size is 95,914,923,278  speedup is 1.00
# umount /koji2
# xfs_repair /dev/stratis/stest/skoji 
Phase 1 - find and verify superblock...
        - reporting progress in intervals of 15 minutes
Phase 2 - using internal log
        - zero log...
        - scan filesystem freespace and inode maps...
        - 18:07:10: scanning filesystem freespace - 32 of 32 allocation groups done
        - found root inode chunk
Phase 3 - for each AG...
        - scan and clear agi unlinked lists...
        - 18:07:10: scanning agi unlinked lists - 32 of 32 allocation groups done
        - process known inodes and perform inode discovery...
        - agno = 0
        - agno = 15
        - agno = 30
        - agno = 16
        - agno = 31
        - agno = 17
        - agno = 18
        - agno = 19
        - agno = 20
        - agno = 21
        - agno = 22
        - agno = 23
        - agno = 24
        - agno = 25
        - agno = 26
        - agno = 27
        - agno = 28
        - agno = 29
        - agno = 1
        - agno = 2
        - agno = 3
        - agno = 4
        - agno = 5
        - agno = 6
        - agno = 7
        - agno = 8
        - agno = 9
        - agno = 10
        - agno = 11
        - agno = 12
        - agno = 13
        - agno = 14
        - 18:07:10: process known inodes and inode discovery - 31360 of 31360 inodes done
        - process newly discovered inodes...
        - 18:07:10: process newly discovered inodes - 32 of 32 allocation groups done
Phase 4 - check for duplicate blocks...
        - setting up duplicate extent list...
        - 18:07:10: setting up duplicate extent list - 32 of 32 allocation groups done
        - check for inodes claiming duplicate blocks...
        - agno = 0
        - agno = 5
        - agno = 1
        - agno = 4
        - agno = 3
        - agno = 6
        - agno = 9
        - agno = 16
        - agno = 8
        - agno = 7
        - agno = 11
        - agno = 10
        - agno = 19
        - agno = 21
        - agno = 13
        - agno = 22
        - agno = 23
        - agno = 24
        - agno = 25
        - agno = 27
        - agno = 15
        - agno = 28
        - agno = 2
        - agno = 17
        - agno = 30
        - agno = 31
        - agno = 20
        - agno = 12
        - agno = 26
        - agno = 29
        - agno = 14
        - agno = 18
        - 18:07:10: check for inodes claiming duplicate blocks - 31360 of 31360 inodes done
Phase 5 - rebuild AG headers and trees...
        - 18:07:10: rebuild AG headers and trees - 32 of 32 allocation groups done
        - reset superblock...
Phase 6 - check inode connectivity...
        - resetting contents of realtime bitmap and summary inodes
        - traversing filesystem ...
        - traversal finished ...
        - moving disconnected inodes to lost+found ...
Phase 7 - verify and correct link counts...
        - 18:07:11: verify and correct link counts - 32 of 32 allocation groups done
xfs_repair: pwrite failed: No space left on device
xfs_repair: pwrite failed: No space left on device
xfs_repair: pwrite failed: No space left on device
xfs_repair: pwrite failed: No space left on device
releasing dirty buffer (bulk) to free list!releasing dirty buffer (bulk) to free list!releasing dirty buffer (bulk) to free list!releasing dirty buffer (bulk) to free list!done
# mount /dev/stratis/stest/skoji /koji2/
mount: /koji2: mount(2) system call failed: Structure needs cleaning.
```

During the whole rsync process there wre quite a few errors in the kernel log eg,
```
.....
Oct 02 18:02:45 d2 kernel: device-mapper: thin: 253:12: switching pool to out-of-data-space (error IO) mode
Oct 02 18:02:45 d2 kernel: xfs_destroy_ioend: 22 callbacks suppressed
Oct 02 18:02:45 d2 kernel: XFS (dm-13): writeback error on sector 104468480
Oct 02 18:02:45 d2 kernel: XFS (dm-13): writeback error on sector 104517632
Oct 02 18:02:45 d2 kernel: XFS (dm-13): writeback error on sector 104574976
Oct 02 18:02:45 d2 kernel: XFS (dm-13): writeback error on sector 100442112
Oct 02 18:02:45 d2 kernel: XFS (dm-13): writeback error on sector 101089280
Oct 02 18:02:45 d2 kernel: XFS (dm-13): writeback error on sector 101736448
Oct 02 18:02:45 d2 kernel: XFS (dm-13): writeback error on sector 102383616
Oct 02 18:02:45 d2 kernel: XFS (dm-13): writeback error on sector 104681472
Oct 02 18:02:45 d2 kernel: device-mapper: thin: 253:12: switching pool to write mode
Oct 02 18:02:45 d2 kernel: device-mapper: thin: 253:12: growing the data device from 45824 to 46848 blocks
...
```

**Comment by yaneti** [comment]
Here is a 
[kernel-stratis-rsync.log](https://github.com/stratis-storage/stratisd/files/2442573/kernel-stratis-rsync.log)
 of all stratis/kernel messages during repeat of the above procedure right after boot of a slightly newer kernel 4.19.0-0.rc6.git1.2.fc30.x86_64

The end result was a xfs system that seemed better (passed xfs_repair) , but had many of the files in it zeroed out.

**Comment by mulkieran** [comment]
Thanks!

**Comment by mulkieran** [comment]
This is the last trouble free extension. Low water mark was hit, stratisd extended, devicemapper extended, all good.

```
Oct 03 17:26:12 d2 kernel: device-mapper: thin: 253:12: reached low water mark for data device: sending event.
Oct 03 17:26:16 d2 kernel: XFS (dm-11): Mounting V5 Filesystem
Oct 03 17:26:16 d2 kernel: XFS (dm-11): Ending clean mount
Oct 03 17:26:16 d2 kernel: XFS (dm-11): Unmounting Filesystem
Oct 03 17:26:16 d2 stratisd[503]:  INFO libstratis::engine::strat_engine::thinpool::thinpool: Attempting to extend thinpool data device belonging to pool db9998fa-e44f-42c0-a3ef-72abdc6a1ade by 2097152 sectors
Oct 03 17:26:16 d2 kernel: device-mapper: thin: 253:12: growing the data device from 768 to 1792 blocks
Oct 03 17:26:16 d2 stratisd[503]:  INFO libstratis::engine::strat_engine::thinpool::thinpool: Extended thin data device by 2097152 sectors
Oct 03 17:26:16 d2 stratisd[503]:  INFO libstratis::engine::strat_engine::thinpool::thinpool: Data tier percent used: 1
```

But pretty soon it got worse, as pool went into out-of-data-space briefly before being grown.

```
Oct 03 17:27:11 d2 kernel: device-mapper: thin: 253:12: reached low water mark for data device: sending event.
Oct 03 17:27:12 d2 stratisd[503]:  INFO libstratis::engine::strat_engine::thinpool::thinpool: Attempting to extend thinpool data device belonging to pool db9998fa-e44f-42c0-a3ef-72abdc6a1ade by 2097152 sectors
Oct 03 17:27:17 d2 kernel: device-mapper: thin: 253:12: switching pool to out-of-data-space (error IO) mode
Oct 03 17:27:18 d2 kernel: device-mapper: thin: 253:12: switching pool to write mode
Oct 03 17:27:18 d2 kernel: device-mapper: thin: 253:12: growing the data device from 1792 to 2816 blocks
Oct 03 17:27:18 d2 stratisd[503]:  INFO libstratis::engine::strat_engine::thinpool::thinpool: Extended thin data device by 2097152 sectors
```

The next time it happened, XFS felt the effects:

```
Oct 03 17:27:20 d2 kernel: device-mapper: thin: 253:12: reached low water mark for data device: sending event.
Oct 03 17:27:22 d2 stratisd[503]:  INFO libstratis::engine::strat_engine::thinpool::thinpool: Attempting to extend thinpool data device belonging to pool db9998fa-e44f-42c0-a3ef-72abdc6a1ade by 2097152 sectors
Oct 03 17:27:26 d2 kernel: XFS (dm-13): writeback error on sector 3241984
Oct 03 17:27:27 d2 kernel: device-mapper: thin: 253:12: switching pool to out-of-data-space (error IO) mode
Oct 03 17:27:27 d2 kernel: XFS (dm-13): writeback error on sector 4382720
Oct 03 17:27:27 d2 kernel: XFS (dm-13): writeback error on sector 4472832
Oct 03 17:27:27 d2 kernel: XFS (dm-13): writeback error on sector 4562944
Oct 03 17:27:27 d2 kernel: XFS (dm-13): writeback error on sector 4696064
Oct 03 17:27:27 d2 kernel: XFS (dm-13): writeback error on sector 4786176
Oct 03 17:27:27 d2 kernel: XFS (dm-13): writeback error on sector 4876288
Oct 03 17:27:27 d2 kernel: XFS (dm-13): writeback error on sector 4966400
Oct 03 17:27:27 d2 kernel: XFS (dm-13): writeback error on sector 5056512
Oct 03 17:27:27 d2 kernel: XFS (dm-13): writeback error on sector 5146624
Oct 03 17:27:28 d2 kernel: device-mapper: thin: 253:12: switching pool to write mode
Oct 03 17:27:28 d2 kernel: device-mapper: thin: 253:12: growing the data device from 2816 to 3840 blocks
Oct 03 17:27:28 d2 stratisd[503]:  INFO libstratis::engine::strat_engine::thinpool::thinpool: Extended thin data device by 2097152 sectors
```

And the last in the log:

```
Oct 03 17:41:09 d2 kernel: device-mapper: thin: 253:12: reached low water mark for data device: sending event.
Oct 03 17:41:18 d2 kernel: device-mapper: thin: 253:12: switching pool to out-of-data-space (error IO) mode
Oct 03 17:41:18 d2 kernel: xfs_destroy_ioend: 1 callbacks suppressed
Oct 03 17:41:18 d2 kernel: XFS (dm-13): writeback error on sector 15968256
Oct 03 17:41:18 d2 kernel: XFS (dm-13): writeback error on sector 201631744
Oct 03 17:41:18 d2 kernel: XFS (dm-13): writeback error on sector 14651392
Oct 03 17:41:22 d2 kernel: XFS (dm-13): writeback error on sector 12670976
Oct 03 17:41:22 d2 kernel: device-mapper: thin: 253:12: switching pool to write mode
Oct 03 17:41:22 d2 stratisd[503]:  INFO libstratis::engine::strat_engine::thinpool::thinpool: Attempting to extend thinpool data device belonging to pool db9998fa-e44f-42c0-a3ef-72abdc6a1ade by 2097152 sectors
Oct 03 17:41:22 d2 kernel: device-mapper: thin: 253:12: growing the data device from 58112 to 59136 blocks
Oct 03 17:41:22 d2 stratisd[503]:  INFO libstratis::engine::strat_engine::thinpool::thinpool: Extended thin data device by 2097152 sectors
Oct 03 17:41:22 d2 stratisd[503]:  INFO libstratis::engine::strat_engine::thinpool::thinpool: Data tier percent used: 38
```

**Comment by yaneti** [comment]
In addition to fixing the core problem here, please consider the error reporting to the upper levels with errors like these.  What I mean is - that rsync run should have never ever succeeded without a peep while the underlying storage was actually failing.

**Comment by mulkieran** [comment]
I'ld like to suggest that this issue be split into a number of distinct bugs. I've already filed a few related bugs.

I think that this issue exposes a few problems:
1. Inconsistent logging. I've already filed one issue about that: #1255.
2. data lowater floor being set too low, so that stratisd's correct reaction is too long delayed. There's a PR for that, #1260. I think it would be good to have a separate issue as well.
3. stratisd's inability to detect bad things that happen that DM doesn't notify it about. In a way, we don't care what happens w/ DM so long as the file system remains in good shape, i.e., it DM goes into out-of-data-space and file system doesn't have a problem then that is not yet a problem, more of a close shave. But, when the file system itself encounters a problem, that is some kind of a failure that should be identified.
